core/include/scipp/core/multi_index.h

```diff
@@ -30,276 +47,332 @@ auto get_slice_dim(const T &param, const Ts &... params) {
+                      std::array<scipp::index, sizeof...(I)> &rewind,
+                      const StridesArgs &... strides) {
+  const bool res = ((strides[dim] == rewind[I] && strides[dim] != 0) && ...);
+  ((rewind[I] = size * strides[dim]), ...);
+  return res;
+}
+
+// non_flattenable_dim is in the storage order of Dimensions & Strides.
+// It is not possible to flatten dimensions outside of the bin-slice dim
+// because they are sliced by that dim and their layout changes depending on
+// the current bin.
+// But the inner dimensions always have the same layout.
+template <class... StridesArgs>
+[[nodiscard]] scipp::index
+flatten_dims(scipp::index *const out_strides, scipp::index *const out_shape,
+             const Dimensions &dims, const scipp::index non_flattenable_dim,
+             const StridesArgs &... strides) {
+  constexpr scipp::index N = sizeof...(StridesArgs);
+  std::array strides_array{std::ref(strides)...};
+  std::array<scipp::index, N> rewind{};
+  scipp::index dim_write = 0;
+  for (scipp::index dim_read = dims.ndim() - 1; dim_read >= 0; --dim_read) {
+    const auto size = dims.size(dim_read);
+    if (dim_read > non_flattenable_dim &&
+        detail::can_be_flattened(dim_read, size, std::make_index_sequence<N>{},
+                                 rewind, strides...)) {
+      out_shape[dim_write - 1] *= size;
+    } else {
+      out_shape[dim_write] = size;
+      for (scipp::index data = 0; data < N; ++data) {
+        out_strides[data + dim_write * N] = strides_array[data].get()[dim_read];
+      }
+      ++dim_write;
+    }
   }
+  return dim_write;
 }
+} // namespace detail
 
 template <scipp::index N> class MultiIndex {
 public:
+  /// Construct without bins.
   template <class... StridesArgs>
-  explicit MultiIndex(const Dimensions &iterDims,
-                      const StridesArgs &... strides)
-      : m_ndim{iterDims.ndim()} {
-    scipp::index d = iterDims.ndim() - 1;
-    for (const auto size : iterDims.shape()) {
-      m_shape[d--] = size;
-      m_end_sentinel *= size;
-    }
-    copy_strides<N>(m_stride, m_ndim, std::index_sequence_for<StridesArgs...>(),
-                    strides...);
+  explicit MultiIndex(const Dimensions &iter_dims,
+                      const StridesArgs &... strides) {
+    const size_t shape_buffer_size =
+        detail::get_shape_buffer_size(iter_dims.ndim());
+    const size_t strides_buffer_size =
+        detail::get_strides_buffer_size<N>(iter_dims.ndim());
+
+    const auto temp_buffer = std::make_unique<scipp::index[]>(
+        strides_buffer_size + shape_buffer_size);
+    scipp::index *const strides_buffer = temp_buffer.get();
+    scipp::index *const shape_buffer = strides_buffer + strides_buffer_size;
+    m_ndim = detail::flatten_dims(strides_buffer, shape_buffer, iter_dims, 0,
+                                  strides...);
+    m_inner_ndim = m_ndim;
+
+    m_buffer =
+        std::make_unique<scipp::index[]>(detail::get_buffer_size<N>(m_ndim));
```

Comment:
Have you tried simply using a large (enough) fixed-size buffer here to avoid the heap allocation? I think we should still try to avoid allocating here (and above), or anywhere in `MultiIndex`.